Rename account to label where appropriate

ryanofsky opened the pull request:
Rename account to label where appropriate

This change only updates strings and adds RPC aliases, but should simplify the implementation of address labels in https://github.com/bitcoin/bitcoin/pull/7729, by getting renaming out of the way and letting that change focus on semantics.

The difference between accounts and labels is that labels apply only to addresses, while accounts apply to both addresses and transactions (transactions have "from" and "to" accounts). The code associating accounts with transactions is clumsy and unreliable so we would like get rid of it.

---

There is a rebased version of #7729 atop this PR at https://github.com/ryanofsky/bitcoin/commits/pr/label, see https://github.com/bitcoin/bitcoin/pull/7729#issuecomment-338417139.

ryanofsky committed:
Rename account to label where appropriate

This change only updates strings and adds RPC aliases, but should simplify the
implementation of address labels in
https://github.com/bitcoin/bitcoin/pull/7729, by getting renaming out of the
way and letting it focus on semantics.

The difference between accounts and labels is that labels apply only to
addresses, while accounts apply to both addresses and transactions
(transactions have "from" and "to" accounts). The code associating accounts
with transactions is clumsy and unreliable so we would like get rid of it.

ryanofsky committed:
Rename wallet_accounts.py test

This is a separate commit because changing the test at the same time as
renaming it breaks git (and github) rename detection.

MeshCollider commented:
I think this should be part of #11497?

ryanofsky commented:
> Concept ACK, I think this should be part of #11497?

I don't mind combining prs though I don't see what advantage it brings in this case. This PR is backwards compatible and can be merged without waiting for #7729.

promag commented on src/qt/paymentserver.cpp:
```diff
@@ -635,9 +635,9 @@ void PaymentServer::fetchPaymentACK(CWallet* wallet, const SendCoinsRecipient& r
-    QString account = tr("Refund from %1").arg(recipient.authenticatedMerchant);
-    std::string strAccount = account.toStdString();
-    std::set<CTxDestination> refundAddresses = wallet->GetAccountAddresses(strAccount);
+    QString label = tr("Refund from %1").arg(recipient.authenticatedMerchant);
```
Only `std::string` is used, so:
```cpp
std::string label = tr("Refund from %1").arg(recipient.authenticatedMerchant).toStdString();
```

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3287,6 +3289,14 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
 }
 
+// Deprecated account RPCs.
+UniValue getaccountaddress(const JSONRPCRequest& request) { return getlabeladdress(request); }
```
Missing `IsDeprecatedRPCEnabled("accounts")` and throw? Same below.

MeshCollider replied:
This doesn't actually use -deprecatedrpc though, that's done in #11497?

promag replied:
I made the question because in the release notes says these are deprecated.

laanwj commented on src/wallet/wallet.h:
```diff
@@ -1141,15 +1141,15 @@ class CReserveKey final : public CReserveScript
 /** 
- * Account information.
- * Stored in wallet with key "acc"+string account name.
+ * Label information.
```
This structure is account-specific. It's going away when account support is dropped, as there is no state per label, it's just an identifier. So I suggest keeping it at the current name.

laanwj commented on src/rpc/protocol.h:
```diff
@@ -76,7 +76,7 @@ enum RPCErrorCode
     RPC_WALLET_ERROR                = -4,  //!< Unspecified problem with wallet (key not found etc.)
     RPC_WALLET_INSUFFICIENT_FUNDS   = -6,  //!< Not enough funds in wallet or account
-    RPC_WALLET_INVALID_ACCOUNT_NAME = -11, //!< Invalid account name
+    RPC_WALLET_INVALID_LABEL_NAME   = -11, //!< Invalid label name
```
I think usually when renaming errors we keep the old name around as an alias - people are processing this directly to RPC wrappers, and old client code would still use the old name.

promag replied:
Just for curiosity, why isn't `RPC_INVALID_PARAMETER` used instead?

laanwj commented on src/rpc/client.cpp:
```diff
@@ -42,12 +42,16 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "settxfee", 0, "amount" },
     { "getreceivedbyaddress", 1, "minconf" },
     { "getreceivedbyaccount", 1, "minconf" },
+    { "getreceivedbylabel", 1, "minconf" },
```
You don't actually introduce these methods in this PR.

laanwj commented:
Concept ACK - needs qa test for the new RPCs.

laanwj commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3287,6 +3289,14 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
+UniValue getaddressesbyaccount(const JSONRPCRequest& request) { return getaddressesbylabel(request); }
+UniValue getreceivedbyaccount(const JSONRPCRequest& request) { return getreceivedbylabel(request); }
+UniValue listreceivedbyaccount(const JSONRPCRequest& request) { return listreceivedbylabel(request); }
+UniValue setaccount(const JSONRPCRequest& request) { return setlabel(request); }
```
This allows for no way for the method to know whether they've been called as labelXX or accountXX, as we might want to implement slightly different behavior for both, that seems important information?

ryanofsky reviewed (commented):
Added 7 commits b105fb9634ad9165552464d0d0ec8f5bb77cc3b0 -> 68b48f6ddde65d4f9134ab85b82f0b48443cc84c ([pr/acct.1](https://github.com/ryanofsky/bitcoin/commits/pr/acct.1) -> [pr/acct.2](https://github.com/ryanofsky/bitcoin/commits/pr/acct.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/acct.1...pr/acct.2))
Squashed 68b48f6ddde65d4f9134ab85b82f0b48443cc84c -> fec503ac115f3de2de31672bc166c8263c8b1317 ([pr/acct.2](https://github.com/ryanofsky/bitcoin/commits/pr/acct.2) -> [pr/acct.3](https://github.com/ryanofsky/bitcoin/commits/pr/acct.3))

> needs qa test for the new RPCs.

Added wallet-labels test in fec503ac115f3de2de31672bc166c8263c8b1317

jonasschnelli commented:
Thanks!
Concept ACK.

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1469,29 +1469,29 @@ UniValue listreceivedbyaccount(const JSONRPCRequest& request)
             "[\n"
             "  {\n"
             "    \"involvesWatchonly\" : true,   (bool) Only returned if imported addresses were involved in transaction\n"
             "    \"account\" : \"accountname\",  (string) The account name of the receiving account\n"
```
Should be updated, since this is an alias for "label". (Same for other occurrences in this file)

ryanofsky replied:
Good catch, updated in 864a83780bc5f8ac3284d4327205e75ca6d0c703.

maflcko reviewed (commented):
Concept ACK fec503ac115f3de2de31672bc166c8263c8b1317

ryanofsky reviewed (commented):
Added 2 commits fec503ac115f3de2de31672bc166c8263c8b1317 -> 864a83780bc5f8ac3284d4327205e75ca6d0c703 ([pr/acct.3](https://github.com/ryanofsky/bitcoin/commits/pr/acct.3) -> [pr/acct.4](https://github.com/ryanofsky/bitcoin/commits/pr/acct.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/acct.3...pr/acct.4))
Squashed 864a83780bc5f8ac3284d4327205e75ca6d0c703 -> 62df1688ef91372efec864c93cdc8968bd7493cb ([pr/acct.4](https://github.com/ryanofsky/bitcoin/commits/pr/acct.4) -> [pr/acct.5](https://github.com/ryanofsky/bitcoin/commits/pr/acct.5))

Sjors commented on doc/release-notes.md:
```diff
@@ -82,9 +82,17 @@ Low-level RPC changes
-
 - `dumpwallet` no longer allows overwriting files. This is a security measure
   as well as prevents dangerous user mistakes.
+- `getnewaddress` and `addmultisigaddress` RPC `account` named parameters have
```
Do you want to keep an `address` alias around for this named parameter as well, along with a deprecation warning? Otherwise it should probably be marked as a breaking change.

ryanofsky replied:
https://github.com/bitcoin/bitcoin/pull/11536#discussion_r149993521

> Do you want to keep an address alias around for this named parameter as well, along with a deprecation warning? Otherwise it should probably be marked as a breaking change.

Thanks. Somehow I missed this comment earlier. Added in 7ae567f87eb474dafd8b3cb1351a0bdfe9e41208

MarcoFalke commented:
Needs rebase

ryanofsky commented:
> Needs rebase

Just so this isn't the last comment., I have been periodically rebasing this PR. The PR is just a big rename so it gets out of date pretty frequently.

ryanofsky reviewed (commented):
Rebased 62df1688ef91372efec864c93cdc8968bd7493cb -> a8e9ac981aeebcd482277115190374227c8be0b8 ([pr/acct.5](https://github.com/ryanofsky/bitcoin/commits/pr/acct.5) -> [pr/acct.6](https://github.com/ryanofsky/bitcoin/commits/pr/acct.6)) due to conflicts.
Rebased a8e9ac981aeebcd482277115190374227c8be0b8 -> 981ac9209629c90377d7586b33dae396017d2398 ([pr/acct.6](https://github.com/ryanofsky/bitcoin/commits/pr/acct.6) -> [pr/acct.7](https://github.com/ryanofsky/bitcoin/commits/pr/acct.7)) due to more conflicts.
Added 1 commit 981ac9209629c90377d7586b33dae396017d2398 -> 7ae567f87eb474dafd8b3cb1351a0bdfe9e41208 ([pr/acct.7](https://github.com/ryanofsky/bitcoin/commits/pr/acct.7) -> [pr/acct.8](https://github.com/ryanofsky/bitcoin/commits/pr/acct.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/acct.7...pr/acct.8)) implementing Sjors suggestion.
Squashed 7ae567f87eb474dafd8b3cb1351a0bdfe9e41208 -> 8ec5890f540e7d96b64824306830abc928864b46 ([pr/acct.8](https://github.com/ryanofsky/bitcoin/commits/pr/acct.8) -> [pr/acct.9](https://github.com/ryanofsky/bitcoin/commits/pr/acct.9))

ryanofsky commented:
Rebased again 8ec5890f540e7d96b64824306830abc928864b46 -> 5bb0a7b79b16649a49c96b88090af051842ea7d4 ([pr/acct.9](https://github.com/ryanofsky/bitcoin/commits/pr/acct.9) -> [pr/acct.10](https://github.com/ryanofsky/bitcoin/commits/pr/acct.10)) due to conflict with #12062.

ryanofsky commented:
Rebased again 5bb0a7b79b16649a49c96b88090af051842ea7d4 -> 1c0f7c7770fbae467db06f7c7e61b1d29dce7771 ([pr/acct.10](https://github.com/ryanofsky/bitcoin/commits/pr/acct.10) -> [pr/acct.11](https://github.com/ryanofsky/bitcoin/commits/pr/acct.11)) due to conflict with #10677.

sipa commented:
utACK 1c0f7c7770fbae467db06f7c7e61b1d29dce7771

ryanofsky commented:
Rebased 1c0f7c7770fbae467db06f7c7e61b1d29dce7771 -> 96363c3f433a4b5970be036aa1cb4ebf36febeab ([pr/acct.11](https://github.com/ryanofsky/bitcoin/commits/pr/acct.11) -> [pr/acct.12](https://github.com/ryanofsky/bitcoin/commits/pr/acct.12)) due to conflicts with #11403.

laanwj commented:
This is not related to segwit wallet, nor a bugfix, so I'm moving the milestone to 0.17.

Sjors commented on doc/release-notes.md:
```diff
@@ -97,6 +97,15 @@ Low-level RPC changes
+- Wallet `getlabeladdress`, `getreceivedbylabel`, `listreceivedbylabel`, and
+  `setlabel` RPCs have been added to replace to be deprecated
+  `getaccountaddress`, `getreceivedbyaccount`, `listreceivedbyaccount`, and
+  `setaccount` RPCs with no differences in behavior.
```
No point in adding `getlabeladdress` and `getreceivedbylabel` if #7729 gets rid of those? In other words, maybe it's better not to rename those methods and leave their deprecation notice?

Sjors reviewed (commented):
Concept ACK. Lightly tested: worked as expected.

Does this PR imply that #7729 must support existing accounts / labels? E.g. it can't switch those over to a new data structure and there's no way to tell if a wallet still needs to be upgraded?

ryanofsky reviewed (commented):
Rebased 96363c3f433a4b5970be036aa1cb4ebf36febeab -> 71621535c75b848cda1a17b991567e361d346050 ([pr/acct.12](https://github.com/ryanofsky/bitcoin/commits/pr/acct.12) -> [pr/acct.13](https://github.com/ryanofsky/bitcoin/commits/pr/acct.13)) due to minor conflict with #12177.

> Does this PR imply that #7729 must support existing accounts / labels? E.g. it can't switch those over to a new data structure and there's no way to tell if a wallet still needs to be upgraded?

No, I don't think this PR implies that. Labels are already used in rpcdump apis and the qt interface, and aren't a new concept. This is just a rename with no change in behavior to use terminology more consistently.

Sjors commented:
@ryanofsky:
> #7729 does add getlabeladdress

It does, but the PR description says "should be removed according to discussion". Maybe that's out of date?

ryanofsky commented:
> It does, but the PR description says "should be removed according to discussion". Maybe that's out of date?

Probably out of date, or could have been a typo. The discussion in that PR has been going on a while and can a little confusing to follow. If you just look at the code changes and commit messages it should make more sense, and also (sorry to keep repeating) my rebased version of #7729 on top of this PR should be even simpler to understand.

Sjors commented:
@ryanofsky I did look at that, but it confused me because it doesn't match the description of the PR. So I suggested that that PR should probably be closed and reopened with an up to date description.

ryanofsky commented:
> I agree that the commit descriptions themselves are quite readable, they just seem out of context.

Would suggest pointing out specific problems or edits you would like to see. It could perhaps be more appropriate to edit the description than open a new PR.

ryanofsky commented:
Rebased 71621535c75b848cda1a17b991567e361d346050 -> 59df27a20a633731506686a477bec10b6a5110d4 ([pr/acct.13](https://github.com/ryanofsky/bitcoin/commits/pr/acct.13) -> [pr/acct.14](https://github.com/ryanofsky/bitcoin/commits/pr/acct.14)) due to conflict in release notes with #12210.

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1538,29 +1538,29 @@ UniValue listreceivedbyaccount(const JSONRPCRequest& request)
+            "    \"account\" : \"accountname\",  (string) DEPRECATED. Backwards compatible alias for label.\n"
+            "    \"amount\" : x.xxx,             (numeric) The total amount received by addresses with this label\n"
             "    \"confirmations\" : n,          (numeric) The number of confirmations of the most recent transaction included\n"
             "    \"label\" : \"label\"           (string) A comment for the address/transaction, if any\n"
```
Should this also be update to:

`The label of the receiving address. The default label is \"\".\n"`

(as above in `listreceivedbyaddress()`)

ryanofsky replied:
> Should this also be update to:

Thanks, updated comment in c1a4f6fc5ea3b6f21413fcb3bc7aa8f67e6ff5e3

jnewbery reviewed (commented):
Tested ACK 59df27a20a633731506686a477bec10b6a5110d4. One comment inline.

Only other comments I'd make are that perhaps it'd be clearer to do all the variable renaming in the first commit and then the file renaming in the second commit. You may also consider renaming the test script to `wallet_labels.py` to conform with #11796.

ryanofsky commented:
Rebased 59df27a20a633731506686a477bec10b6a5110d4 -> 9508f59d293b0f8c67b7b01acc3793f724ab8fe2 ([pr/acct.14](https://github.com/ryanofsky/bitcoin/commits/pr/acct.14) -> [pr/acct.15](https://github.com/ryanofsky/bitcoin/commits/pr/acct.15)) due to conflict with test rename #11774, segwit wallet #11403, addwitness deprecate #12210
Rebased 9508f59d293b0f8c67b7b01acc3793f724ab8fe2 -> 8b9c889702db915eddfff2576cf89389090c2bf7 ([pr/acct.15](https://github.com/ryanofsky/bitcoin/commits/pr/acct.15) -> [pr/acct.16](https://github.com/ryanofsky/bitcoin/commits/pr/acct.16)) due to conflict in release notes.
Rebased 8b9c889702db915eddfff2576cf89389090c2bf7 -> 3d5d5bc03536a8c50ec4778cd599045db2abf24b ([pr/acct.16](https://github.com/ryanofsky/bitcoin/commits/pr/acct.16) -> [pr/acct.17](https://github.com/ryanofsky/bitcoin/commits/pr/acct.17)) due to conflict with #12193.
Added 1 commits 3d5d5bc03536a8c50ec4778cd599045db2abf24b -> c1a4f6fc5ea3b6f21413fcb3bc7aa8f67e6ff5e3 ([pr/acct.17](https://github.com/ryanofsky/bitcoin/commits/pr/acct.17) -> [pr/acct.18](https://github.com/ryanofsky/bitcoin/commits/pr/acct.18), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/acct.17...pr/acct.18))
Squashed c1a4f6fc5ea3b6f21413fcb3bc7aa8f67e6ff5e3 -> 3b2be2aa88823bbb09f12935694af24cf9f292d5 ([pr/acct.18](https://github.com/ryanofsky/bitcoin/commits/pr/acct.18) -> [pr/acct.19](https://github.com/ryanofsky/bitcoin/commits/pr/acct.19))

Sjors commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3836,21 +3838,23 @@ static const CRPCCommand commands[] =
     { "wallet",             "encryptwallet",                    &encryptwallet,                 {"passphrase"} },
-    { "wallet",             "getaccountaddress",                &getaccountaddress,             {"account"} },
+    { "wallet",             "getlabeladdress",                  &getlabeladdress,               {"label"} },
+    { "wallet",             "getaccountaddress",                &getlabeladdress,               {"account"} },
```
Maybe add a comment to point out that this is an alias?

Sjors commented on src/wallet/rpcwallet.cpp:
```diff
@@ -290,14 +290,14 @@ UniValue setaccount(const JSONRPCRequest& request)
             "\nArguments:\n"
-            "1. \"address\"         (string, required) The bitcoin address to be associated with an account.\n"
-            "2. \"account\"         (string, required) The account to assign the address to.\n"
+            "1. \"address\"         (string, required) The bitcoin address to be associated with an label.\n"
```
"a label"

Sjors commented on doc/release-notes.md:
```diff
@@ -62,6 +62,15 @@ RPC changes
+  `setaccount` RPCs with no differences in behavior.
+- Wallet `listreceivedbylabel`, `listreceivedbyaccount` and `listunspent` RPCs
+  add `label` fields to returned JSON objects that previously only had
+  `account` fields.
```
What's the rationale for adding a label field rather than renaming the account field? Because it wasn't deprecated yet?

Sjors reviewed (approved):
Tested ACK 53bafe82.

ryanofsky reviewed (commented):
Added 2 commits 53bafe820e20dfb8192eedfd1ce0c729536de679 -> f0fd2356ad17db395319080b18bc85dcc7d6f43e ([pr/acct.22](https://github.com/ryanofsky/bitcoin/commits/pr/acct.22) -> [pr/acct.23](https://github.com/ryanofsky/bitcoin/commits/pr/acct.23), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/acct.22...pr/acct.23))
Squashed f0fd2356ad17db395319080b18bc85dcc7d6f43e -> 6a25a42d01f9688dbc7cac2e2e5c3bf31555189a ([pr/acct.23](https://github.com/ryanofsky/bitcoin/commits/pr/acct.23) -> [pr/acct.24](https://github.com/ryanofsky/bitcoin/commits/pr/acct.24))
Rebased 6a25a42d01f9688dbc7cac2e2e5c3bf31555189a -> d596f032dcfc7beeed6ad0208ae706e5b999a584 ([pr/acct.24](https://github.com/ryanofsky/bitcoin/commits/pr/acct.24) -> [pr/acct.25](https://github.com/ryanofsky/bitcoin/commits/pr/acct.25)) due to release notes conflict with #11872

Sjors commented:
> Not sure this is the appropriate place to document an alias. 

A generic comment above the table saying that duplicate references to the same function are an alias would be fine.

ryanofsky commented:
> A generic comment above the table saying that duplicate references to the same function are an alias would be fine.

This isn't strictly true, because the both the client and the server can still distinguish between the functions and treat them differently. See `echo`/`echojson` at https://github.com/bitcoin/bitcoin/blob/d42a4fe5aaae60f33a89bde78f21820abefce922/src/rpc/misc.cpp#L472-L473

I'm happy to add a comment but since I'm not sure how this relates to labels or accounts and this PR already touches ~250 lines, maybe we could improve the RPC framework documentation in a new PR.

Sjors commented:
Sounds like documentation (in a different PR) would indeed be useful.

ryanofsky commented:
> Sounds like documentation (in a different PR) would indeed be useful.

See #12700

PierreRochard commented:
Tested ACK d596f032dcfc7beeed6ad0208ae706e5b999a584

ryanofsky commented:
@laanwj what do you think about about merging this change? All it is doing is renaming "account" to "label" so the terms are used more consistently, and then back adding a few "account" aliases to preserve backwards compatibility. Review status:

Tested ACK
- jnewbery
- sjors
- PierreRochard

utACK
- sipa

Concept ACK
- MeshCollider
- laanwj
- jonasschnelli
- MarcoFalke

jnewbery reviewed (commented):
One comment inline.

Why does this PR not add the `getlabel` and `getaddressesbylabel` rpcs?

I still think this would be clearer if you did all the code changes in the first commit, and then the second commit was just the rename `wallet_accounts.py` -> `wallet_labels.py`

ryanofsky commented on doc/release-notes.md:
```diff
@@ -63,6 +63,15 @@ RPC changes
 - The `fundrawtransaction` RPC will reject the previously deprecated `reserveChangeKey` option.
+- Wallet `getnewaddress` and `addmultisigaddress` RPC `account` named
+  parameters have been renamed to `label` with no change in behavior.
+- Wallet `getlabeladdress`, `getreceivedbylabel`, `listreceivedbylabel`, and
```
> Nit: language is perhaps a little clumsy

Updated in 843ee063efc2afe20ef0dbe076bd5989bfb5c59c

ryanofsky reviewed (commented):
Added 1 commits e2966ceee68f88f0b257b43f9c4091ff0c2d5d0f -> 843ee063efc2afe20ef0dbe076bd5989bfb5c59c ([pr/acct.26](https://github.com/ryanofsky/bitcoin/commits/pr/acct.26) -> [pr/acct.27](https://github.com/ryanofsky/bitcoin/commits/pr/acct.27), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/acct.26...pr/acct.27))
Squashed 843ee063efc2afe20ef0dbe076bd5989bfb5c59c -> d2527bd54e994747d9e24043c91d46c7219d46ac ([pr/acct.27](https://github.com/ryanofsky/bitcoin/commits/pr/acct.27) -> [pr/acct.28](https://github.com/ryanofsky/bitcoin/commits/pr/acct.28))

> Why does this PR not add the getlabel and getaddressesbylabel rpcs?

#7729 adds these with different behavior than the existing account rpcs. The idea is for this PR to just clean up stale "account" terminology and for #7729 to add new functionality for making labels work better.

> I still think this would be clearer if you did all the code changes in the first commit, and then the second commit was just the rename wallet_accounts.py -> wallet_labels.py

Sure, moved rename to second commit.

jnewbery commented on test/functional/wallet_labels.py:
```diff
@@ -0,0 +1,206 @@
+    - getaddressesbyaccount
+    - listaddressgroupings
+    - setlabel
+    - sendfrom (with account arguments)
```
This (and the line below) can be updated to `(with label arguments)`

ryanofsky replied:
> This (and the line below) can be updated to (with label arguments)

This is actually intentional. `sendfrom` and `move` are account-specific and deprecated and won't take label arguments. Idea is that we support receiving *at* addresses but not really sending *from* addresses.

jnewbery replied:
ok, makes sense. I was confused by the code lower down in the test case:

```
        # Check that sendfrom label reduces listaccounts balances.
        for i, label in enumerate(labels):
            to_label = labels[(i+1) % len(labels)]
            node.sendfrom(label.name, to_label.receive_address, amount_to_send)
```

jnewbery commented:
Tested ACK d2527bd54e994747d9e24043c91d46c7219d46ac. Thanks for moving the commits around.

Sorry - I've added one more nit in the test file.
